tests/Unit/DataStructures/DataBox/Test_ObservationBox.cpp

```diff
@@ -0,0 +1,128 @@
+  using return_type = double;
+  static constexpr auto function = multiply_by_two;
+  using argument_tags = tmpl::list<Tag0>;
+};
+
+struct Tag2 : db::SimpleTag {
+  using type = double;
+};
+
+struct Tag2Compute : Tag2, db::ComputeTag {
+  using base = Tag2;
+  using return_type = double;
+  static constexpr auto function = multiply_by_two;
+  using argument_tags = tmpl::list<Tag1>;
+};
+
+struct Tag3 : db::SimpleTag {
+  using type = double;
+};
+
+struct Tag3Compute : Tag3, db::ComputeTag {
+  using base = Tag3;
+  using return_type = double;
+  static constexpr auto function = add_all;
+  using argument_tags = tmpl::list<Tag0, Tag1, Tag2>;
+};
+
+struct ScalarVar : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+struct VectorVar : db::SimpleTag {
+  using type = tnsr::I<DataVector, 3, Frame::Inertial>;
+};
+
+struct VarsCompute : db::ComputeTag,
+                     ::Tags::Variables<tmpl::list<ScalarVar, VectorVar>> {
+  using base = ::Tags::Variables<tmpl::list<ScalarVar, VectorVar>>;
+  using return_type = typename base::type;
+  using argument_tags = tmpl::list<>;
+  static void function(
+      const gsl::not_null<::Variables<tmpl::list<ScalarVar, VectorVar>>*>
+          result) {
+    result->initialize(5, 1.0);
+    for (size_t i = 0; i < 3; ++i) {
+      get<VectorVar>(*result).get(i) += static_cast<double>(i) + 1.0;
+    }
+  }
+};
+
+// Test that we can return values from the apply functions and that we can
+// retrieve both the nested DataBox and the ObservationBox via the argument
+// tags.
+struct SubtractNumberAndReturn {
+  using argument_tags =
+      tmpl::list<::Tags::ObservationBox, ::Tags::DataBox, Tag3>;
+
+  template <typename DataBoxType, typename ComputeTagsList>
+  double operator()(const ObservationBox<ComputeTagsList, DataBoxType>& obs_box,
+                    const DataBoxType& box, const double tag3,
+                    const double number) const {
+    CHECK(get<Tag0>(obs_box) == 2.0);
+    CHECK(get<Tag1>(obs_box) == 4.0);
+    CHECK(get<Tag2>(obs_box) == (2.0 * 2.0 * 2.0));
+    CHECK(get<Tag3>(obs_box) == (2.0 + 2.0 * 2.0 + 2.0 * 2.0 * 2.0));
+    CHECK(get<Tag0>(box) == 2.0);
+    CHECK(get<Tag1>(box) == 4.0);
+    return tag3 - number;
+  }
+};
+
+SPECTRE_TEST_CASE("Unit.DataStructures.DataBox.ObservationBox",
+                  "[Unit][DataStructures]") {
+  const auto db_box =
+      db::create<db::AddSimpleTags<Tag0>, db::AddComputeTags<Tag1Compute>>(2.0);
+  const auto obs_box = make_observation_box<
+      db::AddComputeTags<Tag2Compute, Tag3Compute, VarsCompute>>(db_box);
+  CHECK(get<Tag0>(obs_box) == 2.0);
+  CHECK(get<Tag1>(obs_box) == 4.0);
+  CHECK(get<Tag2>(obs_box) == (2.0 * 2.0 * 2.0));
+  CHECK(get<Tag3>(obs_box) == (2.0 + 2.0 * 2.0 + 2.0 * 2.0 * 2.0));
+  CHECK(&db_box == &get<::Tags::DataBox>(obs_box));
+  CHECK(&obs_box == &get<::Tags::ObservationBox>(obs_box));
+
+  CHECK(apply(SubtractNumberAndReturn{}, obs_box, 5.0) ==
+        (get<Tag3>(obs_box) - 5.0));
```

Comment:
Is there a way to test the compute items get evaluated lazily, i.e. they don't get re-evaluated again? That's part of the promise of the ObservationBox, so it would be good to test.

Reply:
Only thing I can think of is having the compute function set a `thread_local` member variable. I'll see if I can get that to work.

Reply:
instead of `bool was_called`, how about a `size_t times_called` so you can `get` twice and check it was called only once